block.go

```diff
@@ -325,13 +353,26 @@ func OpenBlock(logger log.Logger, dir string, pool chunkenc.Pool) (pb *Block, er
 	}
 	closers = append(closers, tr)
 
-	// TODO refactor to set this at block creation time as
-	// that would be the logical place for a block size to be calculated.
-	bs := blockSize(cr, ir, tsr)
-	meta.Stats.NumBytes = bs
-	err = writeMetaFile(logger, dir, meta)
-	if err != nil {
-		level.Warn(logger).Log("msg", "couldn't write the meta file for the block size", "block", dir, "err", err)
+	// Calculate the block size.
+	// This is done at read and not when creating a block to avoid blocks return zero size.
+	// For example opening a block created by an older library version wouldn't include the size information.
+	{
+		// Rewrite the meta only when the sizes has changed.
+		// This check is to allow opening the db in read only mode without
+		// any meta file rewrites when the blocks size hasn't changed.
+		if meta.Stats.NumBytesChunks != cr.Size() ||
```

Comment:
Do we need this complex check? Can we just check `meta.Stats.NumBytesChunk == 0`? Or `meta.Stats.NumBytes != 0` (if we would keep this field) or `meta.Version == 1` (if we would change meta version)

Reply:
Also can we stop rewritting meta at all? We already use the `.Size` anyway.